samples/core/train_until_good/train_until_good.py

```diff
@@ -0,0 +1,88 @@
+# This sample demonstrates continuous training using a train-eval-check recursive loop.
+# The main pipeline trains the initial model and then gradually trains the model some more until the model evaluation metrics are good enough.
+
+import kfp
+from kfp import components
+
+
+chicago_taxi_dataset_op = components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/e3337b8bdcd63636934954e592d4b32c95b49129/components/datasets/Chicago%20Taxi/component.yaml')
+xgboost_train_on_csv_op = components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/567c04c51ff00a1ee525b3458425b17adbe3df61/components/XGBoost/Train/component.yaml')
+xgboost_predict_on_csv_op = components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/567c04c51ff00a1ee525b3458425b17adbe3df61/components/XGBoost/Predict/component.yaml')
+
+pandas_transform_csv_op = components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/6162d55998b176b50267d351241100bb0ee715bc/components/pandas/Transform_DataFrame/in_CSV_format/component.yaml')
+drop_header_op = kfp.components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/02c9638287468c849632cf9f7885b51de4c66f86/components/tables/Remove_header/component.yaml')
+calculate_regression_metrics_from_csv_op = kfp.components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/616542ac0f789914f4eb53438da713dd3004fba4/components/ml_metrics/Calculate_regression_metrics/from_CSV/component.yaml')
+
+
+# This recursive sub-pipeline trains a model, evaluates it, calculates the metrics and checks them.
+# If the model error is too high, then more training is performed until the model is good.
+@kfp.dsl.graph_component
+def train_until_low_error(starting_model, training_data, true_values):
+    # Training
+    model = xgboost_train_on_csv_op(
+        training_data=training_data,
+        starting_model=starting_model,
+        label_column=0,
+        objective='reg:squarederror',
+        num_iterations=50,
+    ).outputs['model']
+
+    # Predicting
+    predictions = xgboost_predict_on_csv_op(
+        data=training_data,
+        model=model,
+        label_column=0,
+    ).output
+
+    # Calculating the regression metrics    
+    metrics_task = calculate_regression_metrics_from_csv_op(
+        true_values=true_values,
+        predicted_values=predictions,
+    )
+
+    # Checking the metrics
+    with kfp.dsl.Condition(metrics_task.outputs['mean_squared_error'] > 0.01):
+        # Training some more
+        train_until_low_error(
+            starting_model=model,
```

Comment:
qq: I see you're using warm starting here. Is part of the reasons that without this caching will step in and mess up the recursion?

Reply:
>Is part of the reasons that without this caching will step in and mess up the recursion?

I do not think so. The caching works great here and it does not know anything about recursion. Caching works on the lower execution level and is not affected by recursion, loops or conditionals.

I'm not sure I fully understand your question. Can you please help me understand it.

The first model training does not use warm start. Then the loop body continues training that model recursively (which is the goal of the sample).

Without continuing from the previous model, the model won't improve, right? If we abandoned the previous model, then why did we train it?

Not every continuous training pipeline will use this pattern. There can be a pipeline that compares production model, a cleanly-trained model and a mixed model and chooses the best one.

Reply:
> Without continuing from the previous model, the model won't improve, right?

Totally agree. Let me elaborate on my question (of course below is just an edge case, I am not suggesting users should do this):

If we hardcode the model uri inside the `train_until_low_error` graph component, and remove that from the function signature of `train_until_low_error`, and we make `xgboost_train_on_csv_op` pick up a model from the hard-coded uri and override that after finish. Then, each time this recursion will see exactly the same component input spec, so, by default the cache is hit and the actual execution will be skipped.

Did I miss something here?

Reply:
> If we hardcode the model uri inside the `train_until_low_error` graph component, and remove that from the function signature of `train_until_low_error`, and we make `xgboost_train_on_csv_op` pick up a model from the hard-coded uri and override that after finish. Then, each time this recursion will see exactly the same component input spec, so, by default the cache is hit and the actual execution will be skipped.

I think this is true. In that case, each "iteration" is exactly the same. Normally, in recursion, the next iteration depends on previous (and as such is different from the previous iteration). But if we remove that data dependency, then all iterations become the same and so the caching system will skip them.

>pick up a model from the hard-coded uri and override that after finish.
This is important part I missed during the first reading.
Yes, the caching system cannot automatically detect that the data behind t he URI has changed. Doing this is not practical. The caching system is only guaranteed to work correctly when the data is immutable and the components are pure.

If the data is expected to change, then the task that downloads the model from the URI should be marked as volatile (`max_cache_staleness = "P0D"`), so that the cache is disabled.

An even better solution is to not overwrite the data behind the URIs. When using the usual data passing methods this cannot happen. Every new model should be written to a separate URI and this new URI should be passed to the next iteration.

Can you please describe a bigger scenario?
I think you are thinking about resolver-type scenarios.
They can be implemented and can even use cache selectively: The code should not overwrite the model itself, but can overwrite a blob *pointing* to that URI. For example `gs://.../models/prod/latest` has content `gs://.../models/prod/a4bf...7ba1`. Then the model downloading can be split into 1) Getting the active prod model URI (volatile) 2) Downloading from that URI. This way, if the model URI has not really changed, the cache would be used successfully. I'm finishing another sample pipeline demonstrating this workflow.

Reply:
> Can you please describe a bigger scenario?

I think the bigger picture is not quite related to this CL, but in general how caching and even orchestration should be implemented, so please view this comment as non-blocking.

As you can see, caching behavior that is based on input/output spec, or component command line args is sensitive to some corner cases where component behavior has some **hidden** dependencies, and we need to rely on users doing the right thing to make sure things work out properly. For caching, if the artifact passed in can be associated with a fingerprint (which is a sort of MLMD property) then I think perhaps we can get rid of this.

Reply:
>I think the bigger picture is not quite related to this CL, but in general how caching and even orchestration should be implemented, so please view this comment as non-blocking.

Thank you. I think these discussions are really insightful.

>caching behavior that is based on input/output spec, or component command line args is sensitive to some corner cases where component behavior has some hidden dependencies

This is true. But I'm not sure this problem can be solved.
You cannot be 100% sure what the user code will do unless you actually run that code.
The component can have non-deterministic behavior, random behavior (bagging, stochastic gradient descent) and can access arbitrary external data (E.g. today's news.). In general it's not possible to predict or prevent this non-deterministic behavior. The best that we could do is to make the containers hermetic and prevent them from accessing any external resources (services, internet, etc). But that would severely limit the usefulness.

>For caching, if the artifact passed in can be associated with a fingerprint (which is a sort of MLMD property) then I think perhaps we can get rid of this.

Here is a subtle, but important aspect: Unlike TFX's artifacts, KFP's intermediate artifacts are isolated and immutable. So the caching guarantees are significantly easier to hold. The components never write to the artifact repository directly and it usually cannot be easily accessed. It's not possible to pass an artifact with custom URI to the KFP component. The only way to add a new artifact is to execute a component which will produce the artifact (which will be stored at the unique immutable location). This seem to make the described problem impossible since the artifact data cannot really change. Without the possibility of data change, there is not much difference between fingerprint and unique id.

Now let's consider a slightly different problem: Caching and components that access external data that can change.
In general, it's not possible to determine the fingerprint of external data and we cannot force the fingerprint to be calculated by every piece of user code getting anything from the internet.
In KFP, the caching behavior is pretty easy: If a downstream component takes an artifact from an upstream component, then the downstream component will always execute if the upstream has executed. The caching is essentially based on the artifact and execution lineage. So, the user either forces the upstream downloader component to re-execute even when all arguments are the same (which in turn re-executed all downstream components) or the user does not force the upstream to re-execute and the cache is used when possible. There user is in control.

There is also a middle option which allows the pipeline author to perform the data check, but still use cache if the data has not changed. I've described a way to make it possible at the end of the previous comment.

So:
* For people with well-behaved pure components the caching just works.
* For people with volatile components, there is a simple way to disable caching for those components.
* There is also a slightly more complicated way to regain some caching support even for the volatile components.

Some action items for myself based on this discussion:

* Add a component that peeks the MD5 checksum of a GCS blob without downloading it
* Add a component that calculates a the data pointed by an HTTP URL
* Add a component that downloads the data from HTTP which optionally accepts the expected data hash. The downloaded data will be checked against the hash.
* Add a component that downloads the data from GCS which optionally accepts the expected data hash. The downloaded data will be checked against the hash.

With this components it would be pretty easy to implement data "peeking" in a cache-preserving way.
First the data hash is calculated. If the hash is the same as before, then the cache will be used downstream. Otherwise, the cache won't be used and the data will be downloaded and processed.